src/connection/mqtt/mod.rs

```diff
@@ -0,0 +1,844 @@
+/*
+ * This file is part of Astarte.
+ *
+ * Copyright 2023 SECO Mind Srl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+pub mod payload;
+
+use std::{
+    collections::HashMap,
+    fmt::{Debug, Display},
+    ops::Deref,
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use log::{debug, error, info, trace, warn};
+use once_cell::sync::OnceCell;
+use rumqttc::{Event as MqttEvent, Packet};
+use serde::{Deserialize, Serialize};
+use tokio::sync::{Mutex, RwLock};
+
+#[cfg(test)]
+pub(crate) use crate::mock::{MockAsyncClient as AsyncClient, MockEventLoop as EventLoop};
+use crate::{
+    builder::{ConnectionConfig, DeviceBuilder, DEFAULT_CHANNEL_SIZE},
+    interface::{
+        mapping::path::MappingPath,
+        reference::{MappingRef, ObjectRef},
+        InterfaceError, Ownership,
+    },
+    interfaces::Interfaces,
+    pairing, properties,
+    retry::DelayedPoll,
+    shared::SharedDevice,
+    store::{error::StoreError, wrapper::StoreWrapper, PropertyStore, StoredProp},
+    topic::ParsedTopic,
+    types::AstarteType,
+    validate::{ValidatedIndividual, ValidatedObject},
+    Interface, Timestamp,
+};
+
+#[cfg(not(test))]
+pub(crate) use rumqttc::{AsyncClient, EventLoop};
+
+use payload::Payload;
+
+use super::{Connection, ReceivedEvent};
+
+/// Borrowing wrapper for the client id
+///
+/// To avoid directly allocating and returning a [`String`] each time
+/// the client id is needed this trait implements [`Display`]
+/// while only borrowing the field needed to construct the client id.
+#[derive(Debug, Clone, Copy)]
+struct ClientId<'a> {
+    realm: &'a str,
+    device_id: &'a str,
+}
+
+impl<'a> Display for ClientId<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}/{}", self.realm, self.device_id)
+    }
+}
+
+/// Shared data of the connection, this struct is internal to the [`Mqtt`] connection
+/// where is wrapped in an arc to share an immutable reference across tasks.
+pub struct SharedMqtt {
+    realm: String,
+    device_id: String,
+    eventloop: Mutex<EventLoop>,
+}
+
+#[derive(Clone)]
+pub struct Mqtt {
```

Comment:
I've opened [an issue referencing this comment](https://github.com/astarte-platform/astarte-device-sdk-rust/issues/257#issue-1990455892) to keep track of this.
Currently I avoided adding the lint to code in this pr to avoid cluttering it too much.